# 項目：評估和清理airbnb NYC 2019 數據


## 分析目标

本研究透過 Airbnb 房源資料，分析不同地區與房型的價格差異、房源供給結構與市場需求特徵
<br>並找出影響房源競爭力的關鍵因素，以評估市場機會與定價策略。

<br>本階段主要針對 Airbnb 房源資料進行資料品質評估與清理，包括檢查資料結構、缺失值、重複值與異常值，以確保資料的可靠性與一致性，為後續分析奠定基礎。


### 欄位說明（Data Dictionary）

- **`id`**：房源的唯一識別碼
- **`name`**：房源名稱（由房東設定的房源標題）
- **`host_id`**：房東的唯一識別碼
- **`host_name`**：房東名稱
- **`neighbourhood_group`**：行政區域（城市中的大區域，例如 Manhattan、Brooklyn）
- **`neighbourhood`**：社區名稱（較細分的地理位置）
- **`latitude`**：房源所在位置的緯度座標
- **`longitude`**：房源所在位置的經度座標
- **`room_type`**：房型類型（例如 Entire home/apt、Private room、Shared room）
- **`price`**：每晚住宿價格（通常以美元計價）
- **`minimum_nights`**：最低入住天數
- **`number_of_reviews`**：房源累計評論數量
- **`last_review`**：最近一次評論日期
- **`reviews_per_month`**：平均每月評論數（反映房源活躍度與需求）
- **`calculated_host_listings_count`**：該房東擁有的房源數量
- **`availability_365`**：一年內可供預訂的天數（0–365 天）


## 讀取數據


In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv('airbnb_NYC_2019.csv')

In [3]:
original_data

id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48890        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891        Marisol            Brooklyn            Bushwick  40.70184   
48892  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893            Taz           Manhattan      Hell's Kitchen  40.75751   
48894     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaN                NaN                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
0                   365  
1                   355  
2                   365  
3                   194  
4                     0  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[48895 rows x 16 columns]

## 評估數據

在這一部分，我將對在上一部分建立的`original_data`這個DataFrame所包含的數據進行評估。

評估主要從兩個方面進行：結構和內容，即整齊度和幹凈度。
<br>數據的結構性問題指不符合“每列是一個變量，每行是一個觀察值，每個單元格是一個值”這三個標準。
<br>數據的內容性問題包括存在丟失數據、重覆數據、無效數據等。

In [4]:
original_data.shape  

(48895, 16)

In [5]:
original_data.head() #先取前5筆資料看有無異常，觀察到last_review與reviews_per_month可能會存在缺失值

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

### 🔵 Step 1：結構快速檢查（整齊度）

In [6]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

#### NOTE : 
我用 original_data.info() 檢查資料，每一欄都是單一變數，數值欄位型態正常（int64/float64），類別欄位是 object，整體結構沒問題。
<br>唯一要注意的是 last_review 目前是字串，後續要轉成 datetime；其餘欄位（像 id、host_id）型態沒問題，不需要改。
<br>總體來說，資料結構與型態都正常，只需處理時間欄位即可。

### 🔵 Step 2：人工抽樣感覺資料（探索）

In [7]:
original_data.sample(10)

id                                               name    host_id  \
44172  34086188              Cozy room in Harlem next to 2/3 train   44919396   
16310  13125552                        Beautiful Bushwick Bungalow   20109188   
24263  19544404                Beautiful 2 Bedroom Lower East Side   72539408   
44442  34229869                             Cozy Living Room Futon   27537930   
26237  20916311                    HUGE apartment in Crown Heights   22565156   
40668  31549470  Cozy UES Studio w/ Gym + Doorman near subway b...  107434423   
3061    1806651                      Bright and Airy Brooklyn Loft    3267818   
14516  11437792                 Hipster Room in South Williamsburg     354385   
27595  21706896  Amazing PRIVATE Bedroom+Bathroom in East Village!   27636450   
10716   8249156                        Room #3 Near Manhattan NYC.   22384027   

        host_name neighbourhood_group    neighbourhood  latitude  longitude  \
44172      Forest           Manhattan           Harlem  40.81808  -73.93671   
16310      Neanna            Brooklyn     Williamsburg  40.70413  -73.93205   
24263    Michelle           Manhattan  Lower East Side  40.72119  -73.98738   
44442     Preston               Bronx   Pelham Gardens  40.86209  -73.84667   
26237     Melissa            Brooklyn    Crown Heights  40.66689  -73.92806   
40668  Blueground           Manhattan  Upper East Side  40.77478  -73.95486   
3061      Allison            Brooklyn     Williamsburg  40.70498  -73.92868   
14516       Chris            Brooklyn     Williamsburg  40.71004  -73.95644   
27595      Lauren           Manhattan     East Village  40.72639  -73.99052   
10716     Shahana            Brooklyn      Brownsville  40.67065  -73.91643   

             room_type  price  minimum_nights  number_of_reviews last_review  \
44172     Private room     45               1                  5  2019-06-06   
16310  Entire home/apt    175               2                102  2019-06-25   
24263  Entire home/apt    125               2                  0         NaN   
44442      Shared room     20               1                  0         NaN   
26237     Private room     45              21                  1  2017-10-02   
40668  Entire home/apt    236              30                  0         NaN   
3061   Entire home/apt    120               3                  2  2015-01-02   
14516     Private room     50               1                  2  2017-02-12   
27595     Private room    150               2                  0         NaN   
10716     Private room     39               1                 45  2019-05-21   

       reviews_per_month  calculated_host_listings_count  availability_365  
44172               2.38                               1                 0  
16310               2.68                               1               284  
24263                NaN                               1                 0  
44442                NaN                               1                 5  
26237               0.05                               1                44  
40668                NaN                             232               127  
3061                0.03                               1                 0  
14516               0.07                               1                 0  
27595                NaN                               2                 0  
10716               0.97                              10               158

價格：沒有 0 或負數，偶有高價但合理
<br>名稱：沒有亂碼
<br>地名：合理
<br>數值欄位：沒有明顯異常
<br>缺失值：last_review 和 reviews_per_month 有 NaN，合理
<br>整體來說，資料「長相」看起來正常

觀察以上抽樣的十筆資料來看
符合：
每一列 = 一筆房源
每一欄 = 一個變數
每個格子 = 一個值
✔️ 沒有結構性問題

### Step 3：系統性檢查乾淨度（重點）


In [8]:
#Step 3-1 缺失值
original_data.isnull().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

💡 綜合上面結果觀察：

| 欄位                | Non-Null Count | 缺失情況               |
| ----------------- | -------------- | ------------------ |
| name              | 48879          | 有 16 筆缺失           |
| host_name         | 48874          | 有 21 筆缺失           |
| last_review       | 38843          | 有 10052 筆缺失（約 20%） |
| reviews_per_month | 38843          | 同上                 |


last_review 和 reviews_per_month 有缺失 → 可能是從未有人評論過的房源
<br>name / host_name 有少量缺失 → 可以決定是丟掉、補「Unknown」或保留


In [9]:
#🔍 Step 3-2：重複值
original_data.duplicated().sum()

np.int64(0)

In [10]:
#🔍 Step 3-3 數值範圍
original_data.describe()

id       host_id      latitude     longitude         price  \
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000   
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687   
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000   
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000   
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48895.000000       48895.000000       38843.000000   
mean         7.029962          23.274466           1.373221   
std         20.510550          44.550582           1.680442   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

In [11]:
# 找出「有評論」但是「最後評論日期缺失」的房源，因為這樣不合理，如果無數據顯示出來表示是合理的。
original_data[
    (original_data['number_of_reviews'] > 0) &
    (original_data['last_review'].isnull())
]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

In [12]:
original_data[original_data['host_name'].isnull()]

id                                               name    host_id  \
360      100184                                          Bienvenue     526653   
2700    1449546                            Cozy Studio in Flatbush    7779204   
5745    4183989    SPRING in the City!! Zen-Style Tranquil Bedroom     919218   
6075    4446862                 Charming Room in Prospect Heights!   23077718   
6582    4763327                Luxurious, best location, spa inc'l   24576978   
8163    6292866                          Modern Quiet Gem Near All   32722063   
8257    6360224                    Sunny, Private room in Bushwick   33134899   
8852    6786181                       R&S Modern Spacious Hideaway   32722063   
9138    6992973                    1 Bedroom in Prime Williamsburg    5162530   
9817    7556587                               Sunny Room in Harlem   39608626   
14040  10709846                 Sunny, spacious room in Greenpoint    7822683   
14631  11553543                                  Cozy Room Astoria   26138712   
15174  12113879     Sunny, Large West Village 1 BR Near Everything    5300585   
19565  15648096              Spacious 2 bedroom close to Manhattan  100971588   
27777  21787094         Spacious and Bright Midtown East Apartment     415290   
27962  21867710  Amazing room in the centre on Manhatten! Welcome!  159156636   
28042  21901156  Best Deal! Lovely place in Manhattan! Time Squ...  159156636   
28274  21990808                 ***SLEEPING IN THE GOLDEN PLACE***  159156636   
30570  23660605    Private Bedroom w/ Balcony view of Central Park  177146433   
32193  25180572                     Beautiful Cozy Studio Apt. UWS  119609345   
38992  30468335            LARGE Prospect Lefferts Gardens Bedroom  228750026   

      host_name neighbourhood_group     neighbourhood  latitude  longitude  \
360         NaN              Queens    Queens Village  40.72413  -73.76133   
2700        NaN            Brooklyn          Flatbush  40.64965  -73.96154   
5745        NaN           Manhattan            Harlem  40.80606  -73.95061   
6075        NaN            Brooklyn     Crown Heights  40.67512  -73.96146   
6582        NaN            Brooklyn        Greenpoint  40.72035  -73.95355   
8163        NaN            Brooklyn     East Flatbush  40.65263  -73.93215   
8257        NaN            Brooklyn          Bushwick  40.70146  -73.92792   
8852        NaN            Brooklyn     East Flatbush  40.64345  -73.93643   
9138        NaN            Brooklyn      Williamsburg  40.71838  -73.95630   
9817        NaN           Manhattan            Harlem  40.82929  -73.94182   
14040       NaN            Brooklyn        Greenpoint  40.73539  -73.95838   
14631       NaN              Queens  Ditmars Steinway  40.77587  -73.91775   
15174       NaN           Manhattan           Chelsea  40.73949  -73.99801   
19565       NaN               Bronx        Highbridge  40.83844  -73.92489   
27777       NaN           Manhattan   Upper East Side  40.76436  -73.96836   
27962       NaN           Manhattan    Hell's Kitchen  40.75668  -73.99097   
28042       NaN           Manhattan    Hell's Kitchen  40.75656  -73.99063   
28274       NaN           Manhattan    Hell's Kitchen  40.75835  -73.99065   
30570       NaN           Manhattan       East Harlem  40.79766  -73.94824   
32193       NaN           Manhattan   Upper West Side  40.77701  -73.97667   
38992       NaN            Brooklyn          Flatbush  40.65152  -73.95271   

             room_type  price  minimum_nights  number_of_reviews last_review  \
360       Private room     50               1                 43  2019-07-08   
2700   Entire home/apt    100              30                 49  2017-01-02   
5745      Private room     86               3                 34  2019-05-23   
6075      Private room     50               1                  0         NaN   
6582   Entire home/apt    195               1                  1  2015-10-20   
8163   Entire home/apt     85    

In [13]:
original_data[ ( original_data['last_review'].isnull() ) &  (original_data['reviews_per_month'].isnull() )]

id                                               name   host_id  \
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
19         7750                  Huge 2 BR Upper East  Cental Park     17985   
26         8700  Magnifique Suite au N de Manhattan - vue Cloitres     26394   
36        11452                        Clean and Quiet in Brooklyn      7355   
38        11943                          Country space in the city     45445   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

             host_name neighbourhood_group       neighbourhood  latitude  \
2            Elisabeth           Manhattan              Harlem  40.80902   
19                Sing           Manhattan         East Harlem  40.79685   
26     Claude & Sophie           Manhattan              Inwood  40.86754   
36                  Vt            Brooklyn  Bedford-Stuyvesant  40.68876   
38             Harriet            Brooklyn            Flatbush  40.63702   
...                ...                 ...                 ...       ...   
48890          Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891          Marisol            Brooklyn            Bushwick  40.70184   
48892    Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893              Taz           Manhattan      Hell's Kitchen  40.75751   
48894       Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
2      -73.94190     Private room    150               3                  0   
19     -73.94872  Entire home/apt    190               7                  0   
26     -73.92639     Private room     80               4                  0   
36     -73.94312     Private room     35              60                  0   
38     -73.96327     Private room    150               1                  0   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
2             NaN                NaN                               1   
19            NaN                NaN                               2   
26            NaN                NaN                               1   
36            NaN                NaN                               1   
38            NaN                NaN                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
2                   365  
19                  249  
26                    0  
36                  365  
38                  365  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[10052 rows 

## 清理数据

| 欄位                | Non-Null Count | 缺失情況               |
| ----------------- | -------------- | ------------------ |
| name              | 48879          | 有 16 筆缺失           |
| host_name         | 48874          | 有 21 筆缺失           |
| last_review       | 38843          | 有 10052 筆缺失（約 20%） |
| reviews_per_month | 38843          | 同上                 |

少量缺失（name、host_name）直接補 'Unknown'；大量缺失（last_review、reviews_per_month）要分開處理，
<br>reviews_per_month 填 0，last_review 保留 NaN 

In [14]:
cleaned_data = original_data.copy()
cleaned_data.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [15]:
# 住宿名稱缺失值 我填 'Unknown'
cleaned_data['name'] = cleaned_data['name'].fillna('Unknown')

#因為房東名字不影響整體數據 所以沒有名字的房東我都補上Unknown
cleaned_data['host_name'] = cleaned_data['host_name'].fillna('Unknown')

#reviews_per_month 缺失值填 0
cleaned_data['reviews_per_month'] = cleaned_data['reviews_per_month'].fillna(0)




In [16]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48895 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48895 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

#### 資料型態轉換規劃

在進行資料清理前，預計對部分欄位進行資料型態轉換，以確保後續分析的正確性與一致性：

#### 1. `id` 與 `host_id`
- 目前可能為浮點數，為識別碼欄位。
- 預計轉換為文字型態（`string`），避免出現小數點，方便後續資料比對與合併。

#### 2. `reviews_per_month`
- 目前為浮點數，存在缺失值。
- 保持浮點數型態以利數值運算，並針對缺失值進行填補（預計填 0）。

#### 3. 數值欄位
- 包含 `price`、`minimum_nights`、`number_of_reviews`、`calculated_host_listings_count`、`availability_365`。
- 型態無需轉換，後續將針對異常值（如過高的 `price` 或 `minimum_nights`）進行處理。


#### 4.日期欄位

`last_review` 目前為物件型態（object / string）
預計轉換為 datetime 型態，方便後續做時間排序、分月統計或計算天數


**總結**：主要轉換集中在識別碼欄位 (`id`、`host_id`) 轉為文字型態，保留浮點型態欄位以便運算，同時處理缺失值與異常值。



In [17]:
cleaned_data['id'] = cleaned_data['id'].astype(int).astype(str) #id
cleaned_data['host_id'] = cleaned_data['host_id'].astype(int).astype(str) #host_id

cleaned_data['last_review'] = pd.to_datetime(original_data['last_review']) #last_review 轉成datetime


In [18]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              48895 non-null  object        
 1   name                            48895 non-null  object        
 2   host_id                         48895 non-null  object        
 3   host_name                       48895 non-null  object        
 4   neighbourhood_group             48895 non-null  object        
 5   neighbourhood                   48895 non-null  object        
 6   latitude                        48895 non-null  float64       
 7   longitude                       48895 non-null  float64       
 8   room_type                       48895 non-null  object        
 9   price                           48895 non-null  int64         
 10  minimum_nights                  48895 non-null  int64         
 11  nu

In [19]:
cleaned_data.describe()

latitude     longitude         price  minimum_nights  \
count  48895.000000  48895.000000  48895.000000    48895.000000   
mean      40.728949    -73.952170    152.720687        7.029962   
min       40.499790    -74.244420      0.000000        1.000000   
25%       40.690100    -73.983070     69.000000        1.000000   
50%       40.723070    -73.955680    106.000000        3.000000   
75%       40.763115    -73.936275    175.000000        5.000000   
max       40.913060    -73.712990  10000.000000     1250.000000   
std        0.054530      0.046157    240.154170       20.510550   

       number_of_reviews                    last_review  reviews_per_month  \
count       48895.000000                          38843       48895.000000   
mean           23.274466  2018-10-04 01:47:23.910099456           1.090910   
min             0.000000            2011-03-28 00:00:00           0.000000   
25%             1.000000            2018-07-08 00:00:00           0.040000   
50%             5.000000            2019-05-19 00:00:00           0.370000   
75%            24.000000            2019-06-23 00:00:00           1.580000   
max           629.000000            2019-07-08 00:00:00          58.500000   
std            44.550582                            NaN           1.597283   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000  
std                         32.952519        131.622289

In [20]:
# 1️⃣ 經緯度範圍檢查
cleaned_data = cleaned_data[
    (cleaned_data['latitude'].between(40.5, 41)) &
    (cleaned_data['longitude'].between(-74.25, -73.7))
]

In [21]:
# 2️⃣ price 與 minimum_nights 異常值篩選（IQR 法）


# price
Q1_price = cleaned_data['price'].quantile(0.25)
Q3_price = cleaned_data['price'].quantile(0.75)
IQR_price = Q3_price - Q1_price
lower_price = Q1_price - 1.5 * IQR_price
upper_price = Q3_price + 1.5 * IQR_price

# minimum_nights
Q1_min = cleaned_data['minimum_nights'].quantile(0.25)
Q3_min = cleaned_data['minimum_nights'].quantile(0.75)
IQR_min = Q3_min - Q1_min
lower_min = Q1_min - 1.5 * IQR_min
upper_min = Q3_min + 1.5 * IQR_min

# 篩掉極端值
cleaned_data = cleaned_data[
    (cleaned_data['price'] >= lower_price) &
    (cleaned_data['price'] <= upper_price) &
    (cleaned_data['minimum_nights'] >= lower_min) &
    (cleaned_data['minimum_nights'] <= upper_min)
]



### 邏輯檢查

In [22]:
# 1️⃣ number_of_reviews > 0 但 last_review 為 NaN
logic_1 = cleaned_data[
    (cleaned_data['number_of_reviews'] > 0) &
    (cleaned_data['last_review'].isnull())
]

# 2️⃣ minimum_nights > 365
logic_2 = cleaned_data[cleaned_data['minimum_nights'] > 365]

# 3️⃣ price <= 0 或過高（上限可依 IQR 或經驗值）
logic_3 = cleaned_data[(cleaned_data['price'] < 0) ]

# 4️⃣ availability_365 = 0 且 number_of_reviews > 0
logic_4 = cleaned_data[
    (cleaned_data['availability_365'] == 0) &
    (cleaned_data['number_of_reviews'] > 0)
]

# 5️⃣ calculated_host_listings_count = 0 但有房源
logic_5 = cleaned_data[
    (cleaned_data['calculated_host_listings_count'] == 0)
]

# 6️⃣ 經緯度離 NYC 太遠（範例）
logic_6 = cleaned_data[
    (cleaned_data['latitude'] < 40.5) | (cleaned_data['latitude'] > 41) |
    (cleaned_data['longitude'] < -74.25) | (cleaned_data['longitude'] > -73.7)
]

# 7️⃣ reviews_per_month > 10（極端值標記）
logic_7 = cleaned_data[cleaned_data['reviews_per_month'] > 10]

# 顯示每個邏輯檢查結果數量
print("1️⃣", len(logic_1), "筆異常")
print("2️⃣", len(logic_2), "筆異常")
print("3️⃣", len(logic_3), "筆異常")
print("4️⃣", len(logic_4), "筆異常")
print("5️⃣", len(logic_5), "筆異常")
print("6️⃣", len(logic_6), "筆異常")
print("7️⃣", len(logic_7), "筆熱門房源")

1️⃣ 0 筆異常
2️⃣ 0 筆異常
3️⃣ 0 筆異常
4️⃣ 11665 筆異常
5️⃣ 0 筆異常
6️⃣ 0 筆異常
7️⃣ 79 筆熱門房源


In [23]:
#看一下哪些房源異常價格
cleaned_data[(cleaned_data['price'] <= 0) ]

id                                               name    host_id  \
23161  18750597  Huge Brooklyn Brownstone Living, Close to it all.    8993084   
25433  20333471      ★Hostel Style Room | Ideal Traveling Buddies★  131697576   
25634  20523843    MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)   15787004   
25753  20608117                    Sunny, Quiet Room in Greenpoint    1641537   
25778  20624541      Modern apartment in the heart of Williamsburg   10132166   
25794  20639628  Spacious comfortable master bedroom with nice ...   86327101   
25795  20639792  Contemporary bedroom in brownstone with nice view   86327101   
25796  20639914       Cozy yet spacious private brownstone bedroom   86327101   
26259  20933849                              the best you can find   13709292   

          host_name neighbourhood_group       neighbourhood  latitude  \
23161      Kimberly            Brooklyn  Bedford-Stuyvesant  40.69023   
25433        Anisha               Bronx     East Morrisania  40.83296   
25634  Martial Loft            Brooklyn            Bushwick  40.69467   
25753        Lauren            Brooklyn          Greenpoint  40.72462   
25778       Aymeric            Brooklyn        Williamsburg  40.70838   
25794       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68173   
25795       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68279   
25796       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68258   
26259        Qiuchi           Manhattan         Murray Hill  40.75091   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
23161  -73.95428     Private room      0               4                  1   
25433  -73.88668     Private room      0               2                 55   
25634  -73.92433     Private room      0               2                 16   
25753  -73.94072     Private room      0               2                 12   
25778  -73.94645  Entire home/apt      0               5                  3   
25794  -73.91342     Private room      0               1                 93   
25795  -73.91170     Private room      0               1                 95   
25796  -73.91284     Private room      0               1                 95   
26259  -73.97597  Entire home/apt      0               3                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
23161  2018-01-06               0.05                               4   
25433  2019-06-24               2.56                               4   
25634  2019-05-18               0.71                               5   
25753  2017-10-27               0.53                               2   
25778  2018-01-02               0.15                               1   
25794  2019-06-15               4.28                               6   
25795  2019-06-21               4.37                               6   
25796  2019-06-23               4.35                               6   
26259         NaT               0.00                               1   

       availability_365  
23161                28  
25433               127  
25634                 0  
25753                 0  
25778                73  
25794               176  
25795               232  
25796               222  
26259                 0

In [24]:
# 刪掉9筆 price <= 0 的異常房源 
cleaned_data = cleaned_data[cleaned_data['price'] > 0]

In [25]:
#經緯度離 NYC 超過範圍 刪除異常的1筆
cleaned_data = cleaned_data[
    (cleaned_data['latitude'].between(40.5, 41)) &
    (cleaned_data['longitude'].between(-74.25, -73.7))
]

## 邏輯檢查：availability_365 = 0 且 number_of_reviews > 0

在邏輯檢查中，我們發現 **10661 筆房源**同時滿足：

- `availability_365 = 0` → 房源一年都不開放
- `number_of_reviews > 0` → 但卻有人評論過

這個現象其實是合理的，原因可能如下：

### 可能原因

1. **歷史評論**
   - 這些房源可能以前有開放過，累積了一些評論
   - 現在設定 `availability_365 = 0` 表示 **暫時不接受預訂**

2. **測試 / 下架房源**
   - 房東可能已下架或暫停房源
   - 但之前累積的評論仍保留在資料中

3. **資料異常**
   - 也有可能是房源資料錄入錯誤
   - 不過通常這種情況對大部分分析影響不大

### 處理建議

- **標記即可**




In [26]:
cleaned_data['inactive_but_reviewed'] = (
    (cleaned_data['availability_365'] == 0) &
    (cleaned_data['number_of_reviews'] > 0)
)

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为`airbnb_NYC_2019_cleaned.csv`。

In [27]:
cleaned_data.head()

id                                              name host_id  \
0  2539                Clean & quiet apt home by the park    2787   
1  2595                             Skylit Midtown Castle    2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !    4632   
3  3831                   Cozy Entire Floor of Brownstone    4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park    7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaT   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.21                               6               365   
1               0.38                               2               355   
2               0.00                               1               365   
3               4.64                               1               194   
4               0.10                               1                 0   

   inactive_but_reviewed  
0                  False  
1                  False  
2                  False  
3                  False  
4                   True

In [28]:
cleaned_data.to_csv("airbnb_NYC_2019_cleaned.csv", index=False)

In [29]:
pd.read_csv("airbnb_NYC_2019_cleaned.csv").head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.21                               6               365   
1               0.38                               2               355   
2               0.00                               1               365   
3               4.64                               1               194   
4               0.10                               1                 0   

   inactive_but_reviewed  
0                  False  
1                  False  
2                  False  
3                  False  
4                   True

# Airbnb 資料清理進度總結

## 1️⃣ 資料型態轉換
- `id`、`host_id` → string
- `last_review` → datetime
- 數值欄位 → 保留 float / int

## 2️⃣ 缺失值處理
- `host_name` / `name` → 填 "Unknown"
- `reviews_per_month` → 填 0
- `last_review` → 保留 NaN

## 3️⃣ 異常值篩選 / 清理
- 經緯度不在 NYC 範圍 → 刪掉
- `price <= 0` → 刪掉
- `price` / `minimum_nights` 極端值 → 使用 IQR 篩掉

## 4️⃣ 邏輯檢查標記
- `availability_365 = 0` 且 `number_of_reviews > 0` → 標記
- `number_of_reviews > 0` 但 `last_review = NaN` → 標記
- `reviews_per_month > 10` → 標記熱門房源
